tpu-client/src/tpu_client.rs

```diff
@@ -95,6 +105,66 @@ where
         self.invoke(self.tpu_client.try_send_transaction(transaction))
     }
 
+    /// Serialize and send transaction to the current and upcoming leader TPUs according to fanout
+    /// size
+    /// Attempts to send and confirm tx "tries" times
+    /// Waits for signature confirmation before returning
+    /// Returns the transaction signature
+    pub fn send_and_confirm_transaction_with_retries<T: Signers + ?Sized>(
+        &self,
+        keypairs: &T,
+        transaction: &mut Transaction,
+        tries: usize,
+        pending_confirmations: usize,
+    ) -> TransportResult<Signature> {
+        for x in 0..tries {
+            let now = Instant::now();
+            let mut num_confirmed = 0;
+            let mut wait_time = MAX_PROCESSING_AGE;
+            // resend the same transaction until the transaction has no chance of succeeding
+            let wire_transaction =
+                bincode::serialize(&transaction).expect("transaction serialization failed");
+
+            while now.elapsed().as_secs() < wait_time as u64 {
```

Comment:
I would define wait_time to be `Duration`. In this case, you don't need `as u64` and also `as_secs`

Reply:
so the only issue here is it does as some extra code as a result. I can do: 
```
let mut wait_time = Duration::from_secs(MAX_PROCESSING_AGE as u64);
...
while now.elapsed() < wait_time {
...
    wait_time = wait_time.max(
        Duration::from_secs((MAX_PROCESSING_AGE * pending_confirmations.saturating_sub(num_confirmed)) as u64));
}
```
^ just makes the wait time a little more convoluted, but if it's generally better to compare durations rather than seconds that is fine with me!

Reply:
right, you are modifying it later. I wonder what is the max total time we can spend retrying

Reply:
ok so `TpuClient::send_and_confirm_transaction_with_retries()` is based off of `ThinClient::send_and_confirm_transaction()`. `pending_confirmations` provides an option to wait for a certain number of blocks to be confirmed before returning. So, when we call `poll_for_signature_confirmation()`, we get back the number of confirmed blocks since the tx signature was confirmed. 

If `confirmed_blocks >= pending_confirmations`, then we are good to go. If not, then we need to wait longer for blocks to progress to reach the `pending_confirmations` threshold. So that is where the `MAX_PROCESSING_AGE * pending_confirmations.saturating_sub(num_confirmed),` comes into play. So if `pending_confirmations=4` and we only get back 1 `confirmed_blocks`. then we need to wait longer for the next three blocks to get confirmed. So we now have to increase the wait time to account for that. 

And yes, this function `TpuClient::send_and_confirm_transaction_with_retries()`, just like the `ThinClient` version is only called in local cluster tests.